pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaAbstractSource.java

```diff
@@ -253,6 +265,12 @@ public void ack() {
             completableFuture.complete(null);
         }
 
+        @Override
+        public void fail() {
+            completableFuture.completeExceptionally(
+               new RuntimeException("Failed to process record with key: " + getKey() + " and value: " + getValue()));
```

Comment:
I wonder if it would be better to skip appending `getValue()` to the exception message?

Reply:
Make sense to me. I used "topic:partition:offset" to replace it, so that a unique message in Kafka can be located.